<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>budden2015treeome by uomsystemsbiology</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>budden2015treeome</h1>
        <p>Code for the manuscript &#39;Modelling the conditional regulatory activity of methylated and bivalent promoters&#39;</p>

        <p class="view"><a href="https://github.com/uomsystemsbiology/budden2015treeome">View the Project on GitHub <small>uomsystemsbiology/budden2015treeome</small></a></p>


        <ul>
          <li><a href="https://github.com/uomsystemsbiology/budden2015treeome/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/uomsystemsbiology/budden2015treeome/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/uomsystemsbiology/budden2015treeome">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a id="modelling-the-conditional-regulatory-activity-of-methylated-and-bivalent-promoters" class="anchor" href="#modelling-the-conditional-regulatory-activity-of-methylated-and-bivalent-promoters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Modelling the conditional regulatory activity of methylated and bivalent promoters</h3>

<p><strong>Motivation</strong>: Predictive modelling of gene expression is a powerful framework for the in silico exploration of transcriptional regulatory interactions through the integration of high-throughput -omics data. A major limitation of previous approaches is their inability to handle conditional and synergistic interactions that emerge when collectively analysing genes subject to different regulatory mechanisms. This limitation reduces overall predictive power and thus the reliability of downstream biological inference.</p>

<p><strong>Results</strong>: We introduce an analytical modelling framework (TREEOME: tree of models of expression) that integrates epigenetic and transcriptomic data by separating genes into putative regulatory classes. Current predictive modelling approaches have found both DNA methylation and histone modification epigenetic data to provide little or no improvement in accuracy of prediction of transcript abundance despite, for example, distinct anti-correlation between mRNA levels and promoter-localised DNA methylation. To improve on this, in TREEOME we evaluate four possible methods of formulating gene-level DNA methylation metrics, which provide a foundation for identifying gene-level methylation events and subsequent differential analysis, whereas most previous techniques operate at the level of individual CpG dinucleotides. We demonstrate TREEOME by integrating gene-level DNA methylation (bisulfite-seq) and histone modification (ChIP-seq) data to accurately predict genome-wide mRNA transcript abundance (RNA-seq) for H1-hESC and GM12878 cell lines. </p>

<h3>
<a id="where-is-the-publication" class="anchor" href="#where-is-the-publication" aria-hidden="true"><span class="octicon octicon-link"></span></a>Where is the publication?</h3>

<p><a href="http://arxiv.org/abs/1502.01409">Here is the arXiv link</a></p>

<h3>
<a id="how-do-i-see-the-results-and-work-with-the-code" class="anchor" href="#how-do-i-see-the-results-and-work-with-the-code" aria-hidden="true"><span class="octicon octicon-link"></span></a>How do I see the results and work with the code?</h3>

<p>We suggest you use the Virtual Reference Environment for the project (what is this?), built using <a href="http://www.vagrantup.com">Vagrant</a>.  To do this, first install <a href="http://www.vagrantup.com">Vagrant</a> and <a href="https://www.virtualbox.org/">VirtualBox</a>, then do:</p>

<pre><code>$ git clone https://github.com/uomsystemsbiology/budden2015treeome.git
$ cd hbgm_vagrant
$ vagrant up
</code></pre>

<p>This will get the code for this project and build a virtual environment with scripts you can run to reproduce the computations and figures in the publication.  </p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/uomsystemsbiology">uomsystemsbiology</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
